include/envoy/server/api_listener.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+namespace Envoy {
+namespace Server {
+
+/**
+ * Handle provided to users to interact with Envoy via a particular API.
+ * Given the flexibility of the api_listener config, this interface is also maximally flexible.
+ * Consumers of this interface have to cast this handle to narrower types.
+ */
+class ApiListenerHandle {
```

Comment:
I like the return patter in address, nice to avoid the dynamic cast. I went ahead and did it. 

I made the ConnectionManagerImpl itself implement Http::ApiListener, but I could also see how it would make sense for the derived HttpApiListener class to implement itself and dispatch the newStream call. There are a couple ways to organize it, let me know what you think.